File: scripts/force_create_missing_functions.py
# force_create_missing_functions.py
# Run in Ghidra (Script Manager or analyzeHeadless) to force-create the 8 functions
# that the OGhidra coverage scan flagged as Ghidra-side gaps.
#
# Source: research/decomp/ogh-coverage-<date>.json -> ghidra_missing_true_gaps
# Plus the 3 Borg-constructor template entries that the prologue scan can't see.
#
# After running: save the project, then re-run ogh-scan-coverage.ps1 — gaps should drop to 0.
#
# Headless invocation:
#   analyzeHeadless.bat D:\GotYaForce\research\decomp GotchaForce ^
#       -process boot.dol ^
#       -scriptPath D:\GotYaForce\scripts ^
#       -postScript force_create_missing_functions.py ^
#       -deleteProject

# @category OGhidra
# @runtime Jython

import json
import os
import re

from ghidra.app.cmd.disassemble import DisassembleCommand
from ghidra.app.cmd.function import CreateFunctionCmd


# Hand-curated template-ctor entries (no prologue; reached via function-pointer table).
# These are the 3 Borg-family constructor dispatchers cited in the port files.
TEMPLATE_CTORS = [
    "0x8007ca5c",  # NORMAL TANK / LEOPARD ctor
    "0x800cfe9c",  # ARROW NINJA / SHIJIMA ctor
    "0x80129608",  # EAGLE ROBOT / PROTO EAGLE ctor
]

# Repo root resolved from this file's location (works in both GUI and headless modes).
SCRIPT_PATH = os.path.dirname(os.path.abspath(__file__))
REPO_ROOT = os.path.dirname(SCRIPT_PATH)


def load_gaps_from_latest_report():
    """Find the newest ogh-coverage-*.json and return its ghidra_missing_true_gaps list."""
    decomp_dir = os.path.join(REPO_ROOT, "research", "decomp")
    if not os.path.isdir(decomp_dir):
        return [], None
    reports = [f for f in os.listdir(decomp_dir) if re.match(r"ogh-coverage-\d{4}-\d{2}-\d{2}\.json$", f)]
    if not reports:
        return [], None
    reports.sort(reverse=True)
    with open(os.path.join(decomp_dir, reports[0])) as f:
        data = json.load(f)
    return data.get("ghidra_missing_true_gaps", []), reports[0]


def force_create(address_hex, monitor):
    """Disassemble + create a function at the given address. Idempotent."""
    addr = currentProgram.getAddressFactory().getAddress(address_hex)
    if addr is None:
        return "skip: bad address"

    fm = currentProgram.getFunctionManager()
    existing = fm.getFunctionAt(addr)
    if existing is not None:
        return "exists: already defined as '%s'" % existing.getName()

    # Clear any data typing at the address, then disassemble.
    listing = currentProgram.getListing()
    cu = listing.getCodeUnitAt(addr)
    if cu is not None and not cu.isInstruction():
        cu.clearCodeUnits()

    DisassembleCommand(addr, None, True).applyTo(currentProgram, monitor)
    CreateFunctionCmd(addr).applyTo(currentProgram, monitor)

    new_fn = fm.getFunctionAt(addr)
    if new_fn is not None:
        return "created: '%s'" % new_fn.getName()
    return "fail: createFunction returned null"


def run():
    gap_list, report_name = load_gaps_from_latest_report()
    print("Loaded %d gaps from %s" % (len(gap_list), report_name or "(no report)"))

    # Merge gap list with hand-curated template ctors (dedup, preserve order).
    all_hex = []
    seen = set()
    for h in gap_list + TEMPLATE_CTORS:
        canonical = h.lower() if h.startswith("0x") else "0x" + h
        if canonical not in seen:
            seen.add(canonical)
            all_hex.append(canonical)

    print("Force-creating %d functions:" % len(all_hex))
    monitor.initialize(len(all_hex))
    created = 0
    skipped = 0
    failed = 0
    for h in all_hex:
        msg = force_create(h, monitor)
        print("  %s : %s" % (h, msg))
        if msg.startswith("created"):
            created += 1
        elif msg.startswith("exists"):
            skipped += 1
        else:
            failed += 1
        monitor.incrementProgress(1)

    print()
    print("Summary: created=%d  existing=%d  failed=%d" % (created, skipped, failed))

    # Save the project so the changes survive a Ghidra restart.
    try:
        currentProgram.save("force-create %d missing functions" % created, monitor)
        print("Program saved.")
    except Exception as e:
        print("WARN: save failed: %s" % str(e))


run()

File: scripts/force_create_missing_functions.java
// force_create_missing_functions.java
// Run in Ghidra (Script Manager or analyzeHeadless) to force-create the 8 functions
// that the OGhidra coverage scan flagged as Ghidra-side gaps.
//
// Source: research/decomp/ogh-coverage-<date>.json -> ghidra_missing_true_gaps
// Plus the 3 Borg-constructor template entries that the prologue scan can't see.
//
// Headless invocation:
//   analyzeHeadless.bat D:\GotYaForce\research\decomp GotchaForce ^
//       -process boot.dol -noanalysis ^
//       -scriptPath D:\GotYaForce\scripts ^
//       -postScript force_create_missing_functions.java
//
// @category OGhidra

import ghidra.app.script.GhidraScript;
import ghidra.app.cmd.disassemble.DisassembleCommand;
import ghidra.app.cmd.function.CreateFunctionCmd;
import ghidra.program.model.address.Address;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.FunctionManager;
import ghidra.program.model.listing.Listing;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;

import java.io.File;
import java.io.FileReader;
import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;
import java.util.regex.Pattern;

public class force_create_missing_functions extends GhidraScript {

    // Hand-curated template-ctor entries (no prologue; reached via function-pointer table).
    private static final String[] TEMPLATE_CTORS = {
        "0x8007ca5c",  // NORMAL TANK / LEOPARD ctor
        "0x800cfe9c",  // ARROW NINJA / SHIJIMA ctor
        "0x80129608",  // EAGLE ROBOT / PROTO EAGLE ctor
    };

    private static final Pattern REPORT_PATTERN =
        Pattern.compile("ogh-coverage-\\d{4}-\\d{2}-\\d{2}\\.json$");

    @Override
    protected void run() throws Exception {
        // Repo root: D:\GotYaForce (script is in D:\GotYaForce\scripts)
        File repoRoot = new File("D:\\GotYaForce");
        if (!repoRoot.exists()) {
            // Fall back: derive from getSourceFile() which returns a ResourceFile
            try {
                Object src = getSourceFile();
                if (src != null) {
                    String path = src.toString();
                    File scriptFile = new File(path);
                    repoRoot = scriptFile.getParentFile().getParentFile();
                }
            } catch (Throwable t) {
                // ignore - use hardcoded path above
            }
        }
        if (repoRoot == null || !repoRoot.exists()) {
            println("ERROR: could not resolve repo root");
            return;
        }

        File decompDir = new File(repoRoot, "research" + File.separator + "decomp");
        List<String> gapList = new ArrayList<>();
        String reportName = null;

        if (decompDir.isDirectory()) {
            File[] reports = decompDir.listFiles(
                (dir, name) -> REPORT_PATTERN.matcher(name).matches());
            if (reports != null && reports.length > 0) {
                File latest = reports[0];
                for (File r : reports) {
                    if (r.getName().compareTo(latest.getName()) > 0) latest = r;
                }
                reportName = latest.getName();
                try (FileReader fr = new FileReader(latest)) {
                    JsonObject root = JsonParser.parseReader(fr).getAsJsonObject();
                    if (root.has("ghidra_missing_true_gaps")) {
                        for (JsonElement e : root.getAsJsonArray("ghidra_missing_true_gaps")) {
                            gapList.add(e.getAsString());
                        }
                    }
                } catch (Exception ex) {
                    println("WARN: failed to parse " + latest + ": " + ex.getMessage());
                }
            }
        }

        println("Loaded " + gapList.size() + " gaps from " +
                (reportName != null ? reportName : "(no report)"));

        // Merge gap list with template ctors (preserve order, dedup, normalize).
        Set<String> merged = new LinkedHashSet<>();
        for (String h : gapList) merged.add(normalize(h));
        for (String h : TEMPLATE_CTORS) merged.add(normalize(h));

        println("Force-creating " + merged.size() + " functions:");
        monitor.initialize(merged.size());
        int created = 0, existing = 0, failed = 0;

        for (String hex : merged) {
            String msg = forceCreate(hex);
            println("  " + hex + " : " + msg);
            if (msg.startsWith("created")) created++;
            else if (msg.startsWith("exists")) existing++;
            else failed++;
            monitor.incrementProgress(1);
            if (monitor.isCancelled()) {
                println("CANCELLED by user");
                break;
            }
        }

        println();
        println("Summary: created=" + created + "  existing=" + existing + "  failed=" + failed);

        try {
            currentProgram.save("force-create " + created + " missing functions", monitor);
            println("Program saved.");
        } catch (Exception ex) {
            println("WARN: save failed: " + ex.getMessage());
        }
    }

    private String normalize(String h) {
        String s = h.toLowerCase();
        if (!s.startsWith("0x")) s = "0x" + s;
        return s;
    }

    private String forceCreate(String addressHex) {
        Address addr = currentProgram.getAddressFactory().getAddress(addressHex);
        if (addr == null) return "skip: bad address";

        FunctionManager fm = currentProgram.getFunctionManager();
        Function existing = fm.getFunctionAt(addr);
        if (existing != null) {
            return "exists: already defined as '" + existing.getName() + "'";
        }

        int tx = currentProgram.startTransaction("forceCreate " + addressHex);
        boolean ok = false;
        try {
            // DisassembleCommand handles clearing any conflicting data typing at the address.
            DisassembleCommand disCmd = new DisassembleCommand(addr, null, true);
            disCmd.applyTo(currentProgram, monitor);

            CreateFunctionCmd fnCmd = new CreateFunctionCmd(addr);
            fnCmd.applyTo(currentProgram, monitor);

            Function newFn = fm.getFunctionAt(addr);
            ok = true;
            return (newFn != null) ? "created: '" + newFn.getName() + "'" : "fail: createFunction returned null";
        } catch (Exception ex) {
            return "fail: " + ex.getMessage();
        } finally {
            currentProgram.endTransaction(tx, ok);
        }
    }
}
